src/Domain/CoordinateMaps/Frustum.cpp

```
@@ -201,7 +211,7 @@ std::optional<std::array<double, 3>> Frustum::inverse(
     logical_coords[2] = (-w_minus_ + w_plus_ * logical_coords[2]) /
                         (w_plus_ - w_minus_ * logical_coords[2]);
   }
-  if (sphericity_ > 0.0) {
+  if (sphericity_ > 0.0 or phi_ != 0.0) {
```

When using this with the interpolation framework, I get an error because `gsl_multiroot` throws. This seems to happen because `block_logical_coordinates` iterates over all blocks and tries to find the block where the map is invertible by checking whether the inverse map returns a value or nullopt. So, I think we should catch this, something like:

```
    try {
      logical_coords = RootFinder::gsl_multiroot(
          rootfunction, logical_coords, absolute_tolerance, maximum_iterations,
          relative_tolerance, verbosity, max_absolute_tolerance, method,
          condition);

    } catch (const convergence_error& e) {
      return std::nullopt;
    }
```
Why your changes make this error show up I do not understand.

---

Is it possible to distinguish between cases where the point is outside the block and hence we _should_ return nullopt, and cases where the root find fails for other reasons and we should throw the error?

---

I doubt there is a good way. One could check the convergence with iterations but that seems very unreliable.

On second thought, what about checking if the flat inverse is in the block before and returning nullopt if it is not?

On third thought, this is what we are already currently doing. So perhaps what has happened here is that the changes introduced in this PR mean that for certain configurations the initial guess / flat inverse will be inside the frustum even though it actually comes from a different block.

---

Hi Niko, I think I need a little help understanding what you encountered so that I know how to update the PR.
What were the parameters of the Frustum map where you saw this error? I am hoping that - if these parameters far from the ones used in the BinaryCompactObject Domain, that I could simply add a comment noting that for particular configurations, the multi-block inversion process is unstable.

Or, if this configuration is in the reasonable range of parameters, could you verify that it is the equiangular option that spoils the multi-block invertibility? If I have a set of parameters for which the frustum is multi-block invertible in the equidistant case and not in the equiangular case then maybe I can use that to narrow down what changes I should make. Thanks!

---

Sure, I think I found a case:

Just taking the same frustum as in  `test_frustum_fail()` in `Test_Frustum:72` and giving it a value that is very far outside the frustum, it will throw a gsl error instead of returning `nullopt`. Can you reproduce this?

```
void test_frustum_fail_equiangular() {
  INFO("Frustum fail");
  const std::array<std::array<double, 2>, 4> face_vertices{
      {{{-2.0, -2.0}}, {{2.0, 2.0}}, {{-4.0, -4.0}}, {{4.0, 4.0}}}};
  const CoordinateMaps::Frustum map(
      face_vertices, 2.0, 5.0, OrientationMap<3>{});
  const std::array<double, 3> test_mapped_point4{{0., 50., 9.0}};
  map.inverse(test_mapped_point4); // works fine
  const CoordinateMaps::Frustum equiangular_map(
      face_vertices, 2.0, 5.0, OrientationMap<3>{}, true, 1.0, false, 1., 1.);
  equiangular_map.inverse(test_mapped_point4); // throws
}
```

Sorry for the long wait!

---

Hi Niko, thanks! I was able to reproduce the error - I think that the map inverse is failing not because of changes in this PR, because the test can be made to pass by reducing the sphericity to a value below `0.525`, independent of what the value for `phi` specified is.

I think the changes that need to be made are to add conditions so that `nullopt` is returned (and so a root find is not attempted) for a larger number of cases, since you found a case where a root find is attempted when it shouldn't be.

Since the map is nonlinear/ not analytically invertible at this point, I don't know if there's a clean way to do this that will catch all cases. I can add a short comment saying that it invertibility is not guaranteed for all Frustum configurations. I can also add a check that if the point to be inverted is outside the radius of the sphere containing the bulged frustum, to return nullopt

I'm not super familiar with how `block_logical_coordinates` uses the inverse map, will these fixes address the issue you ran into? Thanks for looking into this!

---

Hey, 

I think it might be related to the equiangular map because for the `BinaryCompactObject` configuration that I am using this error popped up once I changed it to equiangular even though I was always using sphericity of 1 .

The problem is that `block_logical_coordinates` essentially just tries to invert the point using the inverse map of each block until it works so the inverse must either invert or return `nullopt` but it must not throw.

The simplest thing might be to just wrap the root finding in a `try` block and return nullopt if it fails. However, if you can come up with a way that checks whether the requested point is outside the block and returns `nullopt` in that case that is a lot cleaner of course. If you add this fix I am happy to try it on my `BinaryCompactObject` config.